File: 186.reverse-words-in-a-string-ii/solution.kt
class Solution {
    fun reverseWords(s: CharArray): Unit {
        var L = 0
        var R = s.size - 1
        var tmp: Char = ' '
        val N = s.size

        // Flip the string overall
        while (L < R) {
            tmp = s[L]
            s[L] = s[R]
            s[R] = tmp
            L += 1
            R -= 1
        }

        // Flip each word
        L = 0
        R = 0
        while (R < N) {
            while ((R < N) && (s[R] != ' ')) R++

            val nextR = R + 1
            R -= 1
            while (L < R) {
                tmp = s[L]
                s[L] = s[R]
                s[R] = tmp
                L += 1
                R -= 1
            }
            L = nextR
            R = nextR
        }
    }
}

File: 186.reverse-words-in-a-string-ii/solution.py
class Solution:
    def reverseWords(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        N = len(s)

        # flip the string overall
        L = 0
        R = N - 1
        while (L < R):
            s[L], s[R] = s[R], s[L]
            L += 1
            R -= 1

        # flip character order in each word
        L = R = 0
        while (R < N):
            while (R < N) and (s[R] != " "):
                R += 1

            # flip
            nextR = R
            R -= 1
            while (L < R):
                s[L], s[R] = s[R], s[L]
                L += 1
                R -= 1

            L = R = nextR + 1
        
